File: MCRRTImp/src/main/java/lab/mars/MCRRTImp/model/SimulatedVehicle.java
package lab.mars.MCRRTImp.model;

import lab.mars.RRTBase.Vector;
import lab.mars.RRTBase.Vehicle;

import java.util.ArrayList;
import java.util.List;

public abstract class SimulatedVehicle<V extends Vector<V>> implements Vehicle<V> {

    protected V position;

    protected V velocity;

    protected double rotationLimits;

    protected int numberOfDirection;

    protected double safeDistance;

    public SimulatedVehicle(V position, V velocity, double rotationLimits, int numberOfDirection, double safeDistance) {
        this.position = position;
        this.velocity = velocity;
        this.rotationLimits = rotationLimits;
        this.numberOfDirection = numberOfDirection;
        this.safeDistance = safeDistance;
    }

    protected abstract double simulateVelocity(double currentVelocity, double angle);

    public final void setPosition(V position) {
        this.position = position;
    }

    public final void setVelocity(V velocity) {
        this.velocity = velocity;
    }

    public final V position() {
        return position;
    }

    public final V velocity() {
        return velocity;
    }

    public final double rotationLimits() {
        return rotationLimits;
    }

    public final double safeDistance() {
        return safeDistance;
    }

    public final int numberOfDirection() {
        return numberOfDirection;
    }


    /** Return a RIGHT -> LEFT Point List */
    public final List<Transform<V>> simulateKinetic(V currentVelocity, double deltaTime) {
        List<Transform<V>> ret = new ArrayList<>();
        double deltaTheta = this.rotationLimits / this.numberOfDirection;
        double sliceCount = 100;
        for (int i = -numberOfDirection / 2; i <= numberOfDirection / 2; i++) {
            V rotated = currentVelocity.cpy();
            V translated = currentVelocity.cpy().zero();
            V nextV = currentVelocity.cpy();
            double totalAngleRotated = i * deltaTheta * deltaTime;
            double slicedDeltaTheta = totalAngleRotated / sliceCount;      // The delta for Integrating function
            nextV.rotate(totalAngleRotated);        // Set the Velocity angle
            nextV.normalize().scale(simulateVelocity(currentVelocity.len(), i * deltaTheta));      // Set the Velocity's len()
            double newV = nextV.len();          // Get the Velocity's len()
            for (int c = 0; c < sliceCount; c++) {
                rotated.rotate(slicedDeltaTheta);
                translated.translate(rotated.normalize().scale(newV * deltaTime / sliceCount));
            }
            ret.add(new Transform<>(nextV, translated));
        }
        return ret;
    }


}

File: MCRRTImp/src/main/kotlin/lab/mars/MCRRTImp/model/SimulatedVehicle.kt
package lab.mars.MCRRTImp.model

import lab.mars.MCRRTImp.base.Vector

abstract class SimulatedVehicle<V : Vector<V>>(private val positionV: V, private val velocityV: V, val rotationLimits : Double, val numberOfDirection : Int, val safeDistance : Double) {

    var position: V
        get() = positionV.cpy()
        set(value) {positionV.set(value)}

    var velocity: V
        get() = velocityV.cpy()
        set(value) { velocityV.set(value) }

    protected abstract fun simulateVelocity(currentVelocity : Double, angle : Double) : Double

    open fun simulateKinetic(currentVelocity: V, deltaTime : Double) : List<Transform<V>> {
        val ret = ArrayList<Transform<V>>()
        val deltaTheta = this.rotationLimits / this.numberOfDirection
        val sliceCount = 100.0
        for (i in -numberOfDirection / 2..numberOfDirection / 2) {
            val rotated = currentVelocity.cpy()
            val translated = currentVelocity.cpy().zero()
            val nextV = currentVelocity.cpy()
            val totalAngleRotated = i.toDouble() * deltaTheta * deltaTime
            val slicedDeltaTheta = totalAngleRotated / sliceCount      // The delta for Integrating function
            nextV.rotate(totalAngleRotated)        // Set the Velocity angle
            nextV.normalize().scale(simulateVelocity(currentVelocity.len(), i * deltaTheta))      // Set the Velocity's len()
            val newV = nextV.len()          // Get the Velocity's len()
            var c = 0
            while (c < sliceCount) {
                rotated.rotate(slicedDeltaTheta)
                translated.translate(rotated.normalize().scale(newV * deltaTime / sliceCount))
                c++
            }
            ret.add(Transform(nextV, translated))
        }
        return ret
    }

}